hawkauthlib/utils.py

```diff
@@ -151,12 +151,46 @@ def get_normalized_request_string(request, params=None):
             raise ValueError(msg)
     bits.append(host.lower())
     bits.append(port)
-    bits.append(params.get("hash", ""))
+    bits.append(server_hash if server_hash is not None else "")
     bits.append(params.get("ext", ""))
     bits.append("")     # to get the trailing newline
     return "\n".join(bits)
 
 
+def get_normalized_payload_string(request, params=None):
+    """Get the string to be hashed for Hawk payload verification.
+
+    This function takes a WebOb Request object and returns the data that
+    should be hashed for Hawk payload verification and utilized as the
+    server-side hash calculation for access authentication of that
+    request, a.k.a the "normalized request string".
+
+    If the "params" parameter is not None, it is assumed to be a pre-parsed
+    dict of Hawk parameters as one might find in the Authorization header.  If
+    it is missing or None then the Authorization header from the request will
+    be parsed to determine the necessary parameters.
+    """
+    if not request.text:
+        return None
```

Comment:
> Could we let consumers explicitly opt out of this check in cases where they want to stream the request body?

The spec states that if the hash is not present, you are do not opt in.
The provided code respects this and applies your suggestion already.

> Could we provide an API for allowing the checking of the payload hash after streaming the request body, as a separate operation?

There is a separate method present already, if I understadn you correctly, you're asking to modify it using the bytes stream directly and therefore consuming it in the out-of-band method? This seems to me like it would break many implementations that don't already have a copy (like BytesIO) and are already doing a `seek(0)` first in case someone does something like you're suggesting.

Seems very brittle to me, and makes many assumptions about possible clients mitigation of your suggesting that are not common practice to safely assume are occurring.

I chose to keep with the consistency of the library, we've documented clearly that we're using webob and this would require all of the implementations who consume hawkauthlib to intentionally build the webob request object upfront already (before this change) meaning we can already rely on the `request.text` being prepared for consumption.

The alternative is to add a new dependency on consumers alternatively to the webob we already require, and I was not keen on making that a new requirement.

Reply:
> The spec states that if the hash is not present, you do not opt in at all anyway.

By "consumer" here I meant the server-side code calling into this library; the presence or absence of the `hash` param is a choice made by the client code that's calling the server. I note the mohawk offers a [`accept_untrusted_content` parameter](https://github.com/kumar303/mohawk/blob/b7899166880e890f01cf2531b5686094ba08df8f/mohawk/receiver.py#L45) to let callers opt out of payload verification, which seems similar to what I was trying to get at.

Reply:
> > The spec states that if the hash is not present, you do not opt in at all anyway.
> 
> By "consumer" here I meant the server-side code calling into this library; the presence or absence of the `hash` param is a choice made by the client code that's calling the server. I note the mohawk offers a [`accept_untrusted_content` parameter](https://github.com/kumar303/mohawk/blob/b7899166880e890f01cf2531b5686094ba08df8f/mohawk/receiver.py#L45) to let callers opt out of payload verification, which seems similar to what I was trying to get at.

If the caller opts-out (not provide a hash) then the server has nothing to compare a server-side comparison to.
Also to your point, if the client provided the hash and the server-side is expected to enforce payload verification, then there is a helper I added as mentioned to do this `verify_payload(request, params, algorithm)`.

The server would simply always call `verify_payload()` prior to the normal `check_signature()`. 

If you would like to balance speed with security i'll revert the change to `check_signature` so it uses the client provided hash for speed, and add an `accept_untrusted_content` to allow it to be called the original way :+1: 

Which means this entire PR is no longer considered a 'security patch' i.e. version bump from 2.0.0 to 2.0.0 and it would be a backwards compatibility breaking change with your suggestion, because to have clients continue with current functionality they need to now include `accept_untrusted_content` as a parameter. So a breaking change will version bump 3.0.0 would it not?

My plan is to simply patch the security flaw, i.e. default to accept untrusted content
You're suggesting that it is ok to have `accept_untrusted_content` but to do so we would need to break backwards compatibility.

So should I;

1) add parameter `accept_untrusted_content` to `check_signature()` and revert my `check_signature` security fix, version bump to 3.0.0 according to SemVer. Or;
2) Keep my `check_signature` and version 2.0.1 security fix but document the use `verify_payload()` as a way to optionally enforce payload verification server-side?

Reply:
There's probably a contradiction in option 2, it is more complicated in hawkauthlib compared to mohawk,,, We are relying on `get_signature()` for the actual hash attribute inclusion, the method `check_signature()` alone cannot do what we need.

It becomes complicated _because_ `get_signature()` serves a dual purpose for both server-side and client-side in hawkauthlib.. meaning changing `get_signature()` will cause a circular call for the client-side usage.

I'll push my changes to achieve option 1, as it would be far too large of a refactor to try and treat this as a security patch and support backwards compatibility.. It appears inevitable to make a major version bump and break backwards compatibility to implement the intent of the Hawk Authentication API specification document.

Reply:
>  It appears inevitable to make a major version bump and break backwards compatibility

I agree that's likely the case, and TBH it feels OK even if we *could* do it in a b/w compatible way - this is an important change in functionality that it's worth consumers being explicitly aware of, even if they don't have to change their code to accommodate it.

I don't have any more time to spend on this issue today, but will try to take another look tomorrow and think a bit more about the API surface.

Reply:
I appreciate you attention and glad we made some progress today.

Reply:
Cross-linking for reference, I made some time to think about the API surface here and had some tweaks based on your proposed, which I've pushed in https://github.com/chrisdlangton/hawkauthlib/pull/1. What do you think, does it seem in line with your expectations here?

Reply:
I've updated this PR title and description to capture all our progress.

If the PR you propose is passing it's tests (resolves the circular reference) and it implements the intent of the 3 user journeys in this PR, I'll gladly merge your PR into this.